File: examples/nodejs/hello.mjs
// Minimal walkthrough of the SQLRite Node.js bindings.
//
// Run after:
//
//     cd sdk/nodejs
//     npm install
//     npm run build
//
//     node examples/nodejs/hello.mjs
//
// Shape mirrors `better-sqlite3` so JavaScript devs who've used
// that library can pick this up without reading the docs. Rows
// come back as plain objects keyed by column name.

import { Database } from '../../sdk/nodejs/index.js';

// Pass `:memory:` for a transient in-memory DB (matching better-
// sqlite3's convention); pass a file path like 'foo.sqlrite' for a
// file-backed DB that auto-saves on every write.
const db = new Database(':memory:');

db.exec('CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)');
db.exec("INSERT INTO users (name, age) VALUES ('alice', 30)");
db.exec("INSERT INTO users (name, age) VALUES ('bob', 25)");
db.exec("INSERT INTO users (name, age) VALUES ('carol', 40)");

// .columns() reports projection-order names so you know the shape
// without running the query yet.
const select = db.prepare('SELECT id, name, age FROM users');
console.log('Columns:', select.columns());

// .all() gives you an array of row objects — one object per row.
console.log('\nAll users:');
for (const row of select.all()) {
  console.log(`  ${row.id}: ${row.name} (${row.age})`);
}

// .get() returns the first row (or null if the query is empty) —
// handy for lookups by primary key.
const first = select.get();
console.log('\nFirst user:', first);

// Transactions: BEGIN / INSERT / ROLLBACK leaves the table
// unchanged. The `inTransaction` getter is live throughout.
db.exec('BEGIN');
console.log('\ninTransaction:', db.inTransaction);
db.exec("INSERT INTO users (name, age) VALUES ('phantom', 99)");
const midCount = db.prepare('SELECT id FROM users').all().length;
console.log(`Mid-transaction row count: ${midCount}`);
db.exec('ROLLBACK');
console.log('inTransaction after rollback:', db.inTransaction);
const finalCount = db.prepare('SELECT id FROM users').all().length;
console.log(`Post-rollback row count:   ${finalCount}`);

db.close();

File: examples/python/hello.py
"""Minimal walkthrough of the SQLRite Python bindings.

Run after:

    cd sdk/python
    maturin develop

    python examples/python/hello.py

The shape mirrors the stdlib `sqlite3` module — if you've used that,
you already know how to drive this.
"""

import sqlrite


def main() -> None:
    # Use `:memory:` for a transient in-memory DB (matching sqlite3
    # convention); pass a path like "foo.sqlrite" for a file-backed
    # one that auto-saves on every write.
    with sqlrite.connect(":memory:") as conn:
        cur = conn.cursor()

        cur.execute(
            "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)"
        )
        cur.execute("INSERT INTO users (name, age) VALUES ('alice', 30)")
        cur.execute("INSERT INTO users (name, age) VALUES ('bob', 25)")
        cur.execute("INSERT INTO users (name, age) VALUES ('carol', 40)")

        # `.description` exposes PEP 249 column metadata — a list of
        # 7-tuples, name in position 0, rest None until we track
        # types.
        cur.execute("SELECT id, name, age FROM users")
        print("Columns:", [col[0] for col in cur.description])

        # Iterate tuples DB-API-style.
        print("\nAll users:")
        for row in cur:
            uid, name, age = row
            print(f"  {uid}: {name} ({age})")

        # Transactions: BEGIN + INSERT + ROLLBACK leaves the table
        # unchanged. (`commit()` / `rollback()` work on the
        # Connection; the `with` block auto-commits on clean exit
        # and rolls back on exception.)
        cur.execute("BEGIN")
        cur.execute("INSERT INTO users (name, age) VALUES ('phantom', 99)")
        cur.execute("SELECT id FROM users")
        print(f"\nMid-transaction row count: {len(cur.fetchall())}")

        conn.rollback()
        cur.execute("SELECT id FROM users")
        print(f"Post-rollback row count:   {len(cur.fetchall())}")


if __name__ == "__main__":
    main()
